src/logging.rs

```diff
@@ -0,0 +1,73 @@
+extern crate alloc;
+
+use core::{cell::RefCell, sync::atomic::AtomicUsize};
+
+use alloc::boxed::Box;
+
+use critical_section::RawRestoreState;
+
+struct UartLogger<UartType> {
+    uart: critical_section::Mutex<RefCell<UartType>>,
+    tick_count: Option<&'static AtomicUsize>,
+}
+
+impl<UartType> UartLogger<UartType> {
+    fn new(uart: UartType, tick_count: Option<&'static AtomicUsize>) -> Self {
+        Self {
+            uart: critical_section::Mutex::new(RefCell::new(uart)),
+            tick_count,
+        }
+    }
+}
+
+impl<UartType: Send + core::fmt::Write> log::Log for UartLogger<UartType> {
+    fn enabled(&self, _metadata: &log::Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record) {
+        critical_section::with(|cs| {
+            writeln!(
+                self.uart.borrow(cs).borrow_mut(),
+                "{} {} [{}] {}",
+                record.level(),
+                self.tick_count
+                    .map(|t| (t.load(core::sync::atomic::Ordering::SeqCst) as f32) / 1000.0)
+                    .unwrap_or(0.0),
+                record.module_path().unwrap_or("bin"),
+                record.args()
+            )
+            .ok();
+        })
+    }
+
+    fn flush(&self) {}
+}
+
+
+
+// Dummy critical section code that doesn't really do anything.
+// If we screw up with an interrupt handler, well, oops.
+struct MyCriticalSection;
```

Comment:
We have a critical section [implemented](https://github.com/Foundation-Devices/armv7/blob/update/src/critical_section.rs) in `armv7` crate. It's behind the `critical-section` feature. You can try importing that one instead.